up-streamer/src/ustreamer.rs

```diff
@@ -0,0 +1,974 @@
+/// #             println!(
+/// #                 "UPClientFoo: unregistering topic: {topic:?}"
+/// #             );
+/// #             Ok(())
+/// #         }
+/// #     }
+/// #
+/// #     impl UPClientBar {
+/// #         pub fn new() -> Self {
+/// #             Self {}
+/// #         }
+/// #     }
+/// # }
+/// #
+/// # async fn async_main() {
+///
+/// // Local transport
+/// let local_transport: Arc<dyn UTransport> = Arc::new(up_client_foo::UPClientFoo::new());
+///
+/// // Remote transport router
+/// let remote_transport: Arc<dyn UTransport> = Arc::new(up_client_bar::UPClientBar::new());
+///
+/// // Local endpoint
+/// let local_authority = UAuthority {
+///     name: Some("local".to_string()),
+///     number: Some(Number::Ip(vec![192, 168, 1, 100])),
+///     ..Default::default()
+/// };
+/// let local_endpoint = Endpoint::new("local_endpoint", local_authority, local_transport.clone());
+///
+/// // A remote endpoint
+/// let remote_authority = UAuthority {
+///     name: Some("remote".to_string()),
+///     number: Some(Number::Ip(vec![192, 168, 1, 200])),
+///     ..Default::default()
+/// };
+/// let remote_endpoint = Endpoint::new("remote_endpoint", remote_authority, remote_transport.clone());
+///
+/// let streamer = UStreamer::new("hoge", 100);
+///
+/// // Add forwarding rules to endpoint local<->remote
+/// assert_eq!(
+///     streamer
+///         .add_forwarding_rule(local_endpoint.clone(), remote_endpoint.clone())
+///         .await,
+///     Ok(())
+/// );
+/// assert_eq!(
+///     streamer
+///         .add_forwarding_rule(remote_endpoint.clone(), local_endpoint.clone())
+///         .await,
+///     Ok(())
+/// );
+///
+/// // Add forwarding rules to endpoint local<->local, should report an error
+/// assert!(streamer
+///     .add_forwarding_rule(local_endpoint.clone(), local_endpoint.clone())
+///     .await
+///     .is_err());
+///
+/// // Rule already exists so it should report an error
+/// assert!(streamer
+///     .add_forwarding_rule(local_endpoint.clone(), remote_endpoint.clone())
+///     .await
+///     .is_err());
+///
+/// // Try and remove an invalid rule
+/// assert!(streamer
+///     .delete_forwarding_rule(remote_endpoint.clone(), remote_endpoint.clone())
+///     .await
+///     .is_err());
+///
+/// // remove valid routing rules
+/// assert_eq!(
+///     streamer
+///         .delete_forwarding_rule(local_endpoint.clone(), remote_endpoint.clone())
+///         .await,
+///     Ok(())
+/// );
+/// assert_eq!(
+///     streamer
+///         .delete_forwarding_rule(remote_endpoint.clone(), local_endpoint.clone())
+///         .await,
+///     Ok(())
+/// );
+///
+/// // Try and remove a rule that doesn't exist, should report an error
+/// assert!(streamer
+///     .delete_forwarding_rule(local_endpoint.clone(), remote_endpoint.clone())
+///     .await
+///     .is_err());
+/// # }
+/// ```
+pub struct UStreamer {
+    #[allow(dead_code)]
```

Comment:
why would you need to allow dead code?

Reply:
Was here from when prototyping.

I'll look to see if this is really needed.

Reply:
Turned out this was no longer dead code so just removed the annotation.